h2o-genmodel/src/main/java/hex/genmodel/algos/TreeBasedModel.java

```diff
@@ -0,0 +1,119 @@
+package hex.genmodel.algos;
+
+import hex.genmodel.RawModel;
+import hex.genmodel.utils.ByteBufferWrapper;
+import hex.genmodel.utils.GenmodelBitSet;
+import hex.genmodel.utils.NaSplitDir;
+
+import java.util.Arrays;
+import java.util.Map;
+
+/**
+ * Common ancestor for {@link DrfRawModel} and {@link GbmRawModel}.
+ * See also: `hex.tree.SharedTreeModel` and `hex.tree.TreeVisitor` classes.
+ */
+public abstract class TreeBasedModel extends RawModel {
+    private static final int NsdNaVsRest = NaSplitDir.NAvsREST.value();
+    private static final int NsdNaLeft = NaSplitDir.NALeft.value();
+    private static final int NsdLeft = NaSplitDir.Left.value();
+    protected int _ntrees;
+    protected byte[][] _compressed_trees;
+
+    protected TreeBasedModel(ContentReader cr, Map<String, Object> info, String[] columns, String[][] domains) {
+        super(cr, info, columns, domains);
+        _ntrees = (int) info.get("n_trees");
+    }
+
+    /**
+     * Given a tree (in the form of a byte array) and the row of input data, compute either this tree's
+     * predicted value when `computeLeafAssignment` is false, or the the decision path within the tree (but no more
+     * than 64 levels) when `computeLeafAssignment` is true.
+     *
+     * Note: this function is also used from the `hex.tree.CompressedTree` class in `h2o-algos` project.
+     */
+    @SuppressWarnings("ConstantConditions")  // Complains that the code is too complex. Well duh!
```

Comment:
Regarding comment, we can split it to smaller methods to see if JIT can optimize that in better way ;-) But probably not worth time.